lib/mongoidable/rspec/controller_matchers.rb

```diff
@@ -5,53 +5,156 @@ module RSpec
+
+        attr_reader :controller,
+                    :authorize_action,
+                    :authorized_object
+
+        def loaded_matcher
+          return unless and_load_instance
+
+          expect(controller.instance_variable_get("@#{and_load_instance_name}")).to be
+        end
+
+        def authorization_matcher
+          authorize_callbacks.each do |callback|
+            callback.call(controller)
           end
+
+          matcher = have_received(:authorize!).with(authorize_action, authorized_object)
+          matcher.setup_expectation(controller)
+
+          matcher
         end
 
-        match do |controller|
-          allow(controller).to receive(@controller_action) {}
-          allow(controller.class).to receive(:__callbacks).and_return({})
-          controller.class.define_callbacks :process_action
+        def setup_controller
+          controller.params.merge! controller_params
 
-          CanCan::ControllerResource.add_before_action(controller.class, :authorize_resource, *(@authorize_args || []))
+          controller.params[:action]     = controller_action
+          controller.params[:controller] = controller.class.name.underscore[0..-12]
+          allow(controller).to receive(:action_name).and_return controller_action.to_s
+
+          set_controller_instance_variables
+
+          allow(controller).to receive(:authorize!)
+        end
+
+        def and_load_instance
+          __getobj__.instance_variable_get(:@and_load_instance)
+        end
+
+        def and_load_instance_name
+          __getobj__.instance_variable_get(:@and_load_instance_name) || controller_variable_name
+        end
+
+        def controller_action
+          __getobj__.instance_variable_get(:@controller_action)
+        end
 
-          verifiers = Array.wrap(can_subjects).map do |subject|
-            matcher = ::RSpec::Mocks::Matchers::Receive.new(:authorize!, nil).with(@controller_action, subject)
-            matcher.setup_expectation(controller)
+        def controller_params
+          __getobj__.instance_variable_get(:@controller_params)
+        end
+
+        def after_filter_name
+          __getobj__.instance_variable_get(:@after_filter_name)
+        end
+
+        def set_controller_instance_variables
+          return if instance_variables.blank?
+
+          instance_variables.each do |variable_name, use_value|
+            variable_name ||= controller_variable_name
+            controller.instance_variable_set("@#{variable_name}", use_value)
+          end
+        end
+
+        def instance_variables
+          __getobj__.instance_variable_get(:@instance_variables)
+        end
+
+        def controller_variable_name
+          controller.class.name[0..-11].demodulize.singularize.underscore
+        end
+
+        def authorize_callbacks
+          found_filter = after_filter_name.blank?
+
+          @authorize_callbacks ||= controller._process_action_callbacks.each_with_object([]) do |action, array|
```

Comment:
This is similar to what I originally had. I moved away from this for performance reasons. I was trying to avoid having to loop every filter every time. This could be a spec perf issue. I understand the reason you changed, just wondering if we can come up with a stub instead.

Reply:
So, I don't like my changes, but I couldn't figure out a good stub option either.

I suspect that you and I may be looking at this particular bit of code differently and I think we should talk about it.